Userland/DevTools/UserspaceEmulator/MallocTracer.cpp

```diff
@@ -349,19 +354,51 @@ bool MallocTracer::is_reachable(const Mallocation& mallocation) const
             return IterationDecision::Continue;
 
         size_t pointers_in_region = region.size() / sizeof(u32);
+
         for (size_t i = 0; i < pointers_in_region; ++i) {
             auto value = region.read32(i * sizeof(u32));
-            if (value.value() == mallocation.address && !value.is_uninitialized()) {
+            auto other_address = value.value();
+            if (!value.is_uninitialized() && m_memory_graph.contains(value.value())) {
 #if REACHABLE_DEBUG
-                reportln("mallocation {:p} is reachable from region {:p}-{:p}", mallocation.address, region.base(), region.end() - 1);
+                reportln("region/mallocation {:p} is reachable from region {:p}-{:p}", other_address, region.base(), region.end() - 1);
 #endif
-                reachable = true;
-                return IterationDecision::Break;
+                m_memory_graph.find(other_address)->value.is_reachable = true;
+                reachable_mallocations.append(other_address);
             }
         }
         return IterationDecision::Continue;
     });
-    return reachable;
+
+    // Propagate reachability
+    // There are probably better ways to do that
+    Vector<FlatPtr> visited = {};
+    for (size_t i = 0; i < reachable_mallocations.size(); ++i) {
+        auto reachable = reachable_mallocations.at(i);
+        if (visited.contains_slow(reachable))
+            continue;
+        visited.append(reachable);
+        auto& mallocation_node = m_memory_graph.find(reachable)->value;
+
+        if (!mallocation_node.is_reachable) {
+            mallocation_node.is_reachable = true;
+            m_memory_graph.set(reachable, mallocation_node);
```

Comment:
This set here is also unnecessary since your mallocation node is a reference to a GraphNode in the memory graph.

Reply:
And actually, removing the set makes it more clear that this is almost the exact same thing you're doing on line 365.

A future change set might make `MemoryGraph::set_reachable(FlatPtr, bool)` or similar a thing that should exist.

Could also iterate on this all day and do optimizations such as:
- Make `reachable_mallocations` a vector of `GraphNode*`, since we know all the mallocations will be in the hashmap and have stable addresses. This would let you avoid the hash table lookup when walking your ever-growing reachable_mallocations vector. Would require storing the address in the graph node too, though.
- Make the malloc tracer use the a graph/hashmap structure for all mallocation tracking, not just at report generation time, which would let you avoid the first pass through for_all_mallocations (dubious claim from me, because you are filtering out freed mallocations as well)
- Give reachable_mallocations a quick `reserve` to the number of  nodes in the memory graph, since we know it will need to be at least that big by the end. Same with visited. Maybe number of nodes * some arbitrary scale factor? (1.2? 1.4?)

Buuuut with the two changes I mentioned already it's good enough to merge probably, with the optimizations above to be investigated later :)